grpc/src/tx.rs

```diff
@@ -0,0 +1,602 @@
+    ///     to_address: "celestia169s50psyj2f4la9a2235329xz7rk6c53zhw9mm".to_string(),
+    ///     amount: vec![Coin::utia(12345).into()],
+    /// };
+    ///
+    /// tx_client
+    ///     .submit_message(msg.clone(), TxConfig::default())
+    ///     .await
+    ///     .unwrap();
+    /// # }
+    /// ```
+    pub async fn submit_message<M>(&self, message: M, cfg: TxConfig) -> Result<TxInfo>
+    where
+        M: Name,
+    {
+        let tx_body = RawTxBody {
+            messages: vec![into_any(message)],
+            ..RawTxBody::default()
+        };
+
+        let mut is_retry = false;
+        let (tx_hash, sequence) = loop {
+            match self.sign_and_broadcast_tx(tx_body.clone(), cfg).await {
+                Ok(resp) => break resp,
+                Err(e) if !is_retry => {
+                    if self.maybe_update_gas_price(&e, cfg)? {
+                        is_retry = true;
+                        continue;
+                    }
+                    return Err(e);
+                }
+                Err(e) => return Err(e),
+            }
+        };
+        self.confirm_tx(tx_hash, sequence).await
+    }
+
+    /// Submit given blobs to celestia network.
+    ///
+    /// When no gas price is specified through config, it will automatically
+    /// handle updating client's gas price when consensus updates minimal
+    /// gas price.
+    ///
+    /// # Example
+    /// ```no_run
+    /// # async fn docs() {
+    /// use celestia_grpc::{TxClient, TxConfig};
+    /// use celestia_types::state::{AccAddress, Coin};
+    /// use celestia_types::{AppVersion, Blob};
+    /// use celestia_types::nmt::Namespace;
+    /// use tendermint::crypto::default::ecdsa_secp256k1::SigningKey;
+    ///
+    /// let signing_key = SigningKey::random(&mut rand_core::OsRng);
+    /// let public_key = *signing_key.verifying_key();
+    /// let address = AccAddress::new(public_key.into()).into();
+    /// let grpc_url = "public-celestia-mocha4-consensus.numia.xyz:9090";
+    ///
+    /// let tx_client = TxClient::with_url(grpc_url, &address, public_key, signing_key)
+    ///     .await
+    ///     .unwrap();
+    ///
+    /// let ns = Namespace::new_v0(b"abcd").unwrap();
+    /// let blob = Blob::new(ns, "some data".into(), AppVersion::V3).unwrap();
+    ///
+    /// tx_client
+    ///     .submit_blobs(&[blob], TxConfig::default())
+    ///     .await
+    ///     .unwrap();
+    /// # }
+    /// ```
+    pub async fn submit_blobs(&self, blobs: &[Blob], cfg: TxConfig) -> Result<TxInfo> {
+        if blobs.is_empty() {
+            return Err(Error::TxEmptyBlobList);
+        }
+        for blob in blobs {
+            blob.validate(self.app_version)?;
+        }
+
+        let mut is_retry = false;
+        let (tx_hash, sequence) = loop {
+            match self.sign_and_broadcast_blobs(blobs.to_vec(), cfg).await {
+                Ok(resp) => break resp,
+                Err(e) if !is_retry => {
```

Comment:
What if it fails again? I think we should have more retries. Like 3-5 retries.

Reply:
gas price of the clien is the minimal possible gas price that the network agreed on. My reasoning was that validator voting for a new minimal gas price won't happen as quick as our update and resend, so we are safe with just one retry. I can update it to, say 3

Reply:
done